filebeat/inputsource/common/dgram/server.go

```diff
@@ -72,26 +72,29 @@ func (l *Listener) Run(ctx context.Context) error {
 	l.log.Info("Started listening for " + l.family.String() + " connection")
 
 	for ctx.Err() == nil {
-		conn, err := l.listener()
-		if err != nil {
-			l.log.Debugw("Cannot connect", "error", err)
-			continue
-		}
-		connCtx, connCancel := ctxtool.WithFunc(ctx, func() {
-			conn.Close()
-		})
-
-		err = l.run(connCtx, conn)
-		if err != nil {
-			l.log.Debugw("Error while processing input", "error", err)
-			connCancel()
-			continue
-		}
-		connCancel()
+		l.doRun(ctx)
```

Comment:
moved the body into a separate method, so we can do `defer cancel()` to cleanup resources. The handle is created independent of the module and we must cleanup in case the handler panics.

TODO: shall we consider to recover from a panic here or only recover if the handler failed?

Reply:
Looking at the package `net`, we probably won't see many panics from `listener` function. Also, how could we recover from such issues? If anything fails during that function call it requires action from the user, e.g. fixing addresses or changing the protocol. These are problems from those we cannot recover without user interference.

I would only call recover to take care of panics from `handler`.

Reply:
> I would only call recover to take care of panics from handler.

yeah, I did mean the handler. Added `Recover` to connectAndRun